Paper: Bayesian Estimation and Forecasting of Time Series in statsmodels

ChadFulton opened the pull request:
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.


*Editor*: <!--editor-->@stargaser<!--end-editor-->

*Reviewers*: <!--reviewers-list-->@sidsriva, @Nickledave<!--end-reviewers-list-->

ChadFulton committed:
ENH: initial draft

scoobies commented:
Hi!, I'm @scoobies, a friendly bot.

Type ```@scoobies help``` to discover how I can help you.

ChadFulton committed:
ENH: improve figures

ChadFulton committed:
ENH: Note that PyMC3 can be used with statsmodels

ChadFulton committed:
ENH: Add prophet, causalimpact references in STS

ChadFulton committed:
ENH: update GS example to use multiple parameters

ChadFulton committed:
BUG: remove reference from example paper

deniederhut commented:
@scoobies mark paper

deniederhut commented:
@scoobies build status

scoobies commented:
{"ChadFulton-paper-96"=>{"data"=>{"build_output"=>"[*] Updating proceedings build tools in {self.master_repo_path}...\nAlready up to date.\n[*] Check out paper repository...\nCloning into '/tmp/tmpjxciu0wm'...\nMoving proceedings build tools to temp directory...\n[*] Build the paper...\nBuilding: chad_fulton\n*** Warning: /tmp/tmpjxciu0wm/publisher/../output/chad_fulton/paper_stats.json does not exist.\n", "build_pdf_path"=>"/procbuild/procbuild/../cache/ChadFulton-paper-96.pdf", "build_status"=>"success", "build_timestamp"=>"28/05 07:04"}, "status"=>"success"}}

deniederhut commented:
@scoobies mark needs review

deniederhut commented:
@scoobies assign @stargaser as editor

scoobies commented:
Assigned! @stargaser is now the editor

mepa commented:
@scoobies add @sidsriva to reviewers

scoobies commented:
@sidsriva added to the reviewers list!

deniederhut commented:
@scoobies add @Nickledave to reviewers

scoobies commented:
@Nickledave added to the reviewers list!

deniederhut commented:
Hi! We've moved the deadline for first reviews back one week, to June 16th, to give us a little more time to find reviewers for all the papers this year. You can see the updated full timeline at https://github.com/scipy-conference/scipy_proceedings#timeline-for-2022.

ChadFulton committed:
ENH: add illustrative examples

NickleDave commented on papers/chad_fulton/chad_fulton.rst:
```diff
@@ -0,0 +1,709 @@
+equations that each describe some aspect of the evolution of univariate
+time series data. While originally somewhat *ad hoc*, these models can be
+defined in terms of a proper statistical model (for example, see
+:cite:`hyndman_forecasting_2008`), and, regardless, they have enjoyed
```
This sentence runs a bit long, suggest splitting into two by removing `,and, regardless`

`(for example, see :cite:`hyndman_forecasting_2008`). They have enjoyed`

ChadFulton replied:
Thank you - fixed in current version.

NickleDave commented on papers/chad_fulton/chad_fulton.rst:
```diff
@@ -0,0 +1,709 @@
+time series data. While originally somewhat *ad hoc*, these models can be
+defined in terms of a proper statistical model (for example, see
+:cite:`hyndman_forecasting_2008`), and, regardless, they have enjoyed
+considerably popularity in forecasting (for example, see the implementation in
```
should be "considerable" not "considerably" I think?

ChadFulton replied:
Thank you - fixed in current version.

NickleDave commented on papers/chad_fulton/chad_fulton.rst:
```diff
@@ -0,0 +1,709 @@
+:math:`s_t` is the seasonal component of period :math:`m`, and
+:math:`\alpha, \beta, \gamma` are parameters of the model. When augmented with
+an error term with some given probability distribution (usually Gaussian),
+likelihood-based inference can be used to, for example, estimate the parameters.
```
For me the interjection here ", for example," distracts from reading the sentence. Yes likelihood-based inference can be used for other things but those other things aren't mentioned and it's not adding to the overall point of the paragraph

ChadFulton replied:
Thank you - fixed in current version.

NickleDave commented on papers/chad_fulton/chad_fulton.rst:
```diff
@@ -0,0 +1,709 @@
+specifications for the trend, seasonal, and cyclical components, so that a wide
+variety of time series characteristics can be accommodated. In *statsmodels*,
+these models can be constructed from the :code:`UnobservedComponents` class; a
+few examples are given in the following code
```
Add a colon here?
"following code:"

ChadFulton replied:
Thank you - fixed in current version.

NickleDave commented on papers/chad_fulton/chad_fulton.rst:
```diff
@@ -0,0 +1,709 @@
+
+With either Metropolis-Hastings or Gibbs sampling, our procedure will produce a
+sequence of sample values (of parameters and / or the unobserved state vector)
+that – assuming certain conditions are met – approximate draws from the
```
this interjection with hyphens makes the sentence hard to follow, esp with a verb like "approximate" that can also read as an adjective. Suggest removing, your readers know that stats math usually assumes some assumptions are met

ChadFulton replied:
Thank you - fixed in current version.

NickleDave commented on papers/chad_fulton/chad_fulton.rst:
```diff
@@ -0,0 +1,709 @@
+'''''''''''''''''''''''''''''''
+
+In random walk Metropolis-Hastings (MH), we begin with an arbitrary point as the
+initial sample, and then interatively construct new samples in the chain as
```
"iteratively" not "interatively" (the extra 'n' is a typo)

ChadFulton replied:
Thank you - fixed in current version.

NickleDave commented on papers/chad_fulton/chad_fulton.rst:
```diff
@@ -0,0 +1,709 @@
+follows. At each iteration, (a) construct a proposal by perturbing the previous
+sample by a Gaussian random variable, and then (b) accept the proposal with some
+probability. If a proposal is accepted, it becomes the next sample in the chain,
+while if it is rejected then previous sample value is carried over. Here, we
```
maybe "then the previous" instead of "then previous"?

ChadFulton replied:
Thank you - fixed in current version.

NickleDave commented on papers/chad_fulton/chad_fulton.rst:
```diff
@@ -0,0 +1,709 @@
+distributions of every variable, even though it is still not possible to derive
+the general form of the joint posterior. While this approach can be superior to
+random walk MH when it is applicable, the ability to derive the conditional
+distributions typically requires the use of a "conjugate" prior – i.e. a prior
```
Suggest adding a comma after "i.e." (that is standard, for US style at least https://english.stackexchange.com/questions/6894/should-there-be-a-comma-after-i-e)

ChadFulton replied:
Thank you - fixed in current version.

NickleDave commented:
Hi @ChadFulton thank you for a very clear paper that was a joy to read.  

I have made some minor formatting suggestions that you may consider integrating.  

One point I did not comment on at a line level:  
I see you wrote library and variable names in italics in most places.  
I'm looking at the rendered pdf here:
https://procbuild.scipy.org/download/ChadFulton-paper-96
Is that intentional or did you mean those to appear as monospace?  
(I always forget that in rst you need to use two backticks, unlike markdown that only uses one for monospace, which is why I ask.)

I may go through the end of the paper one more time and suggest a little more. But overall, LGTM!  

I am definitely not a subject matter expert here so I may have missed, e.g., math notation typos. 
In spite of my lack of expertise, I could follow most of the logical flow, something I appreciate.

Will be useful for `statsmodels` users that want to apply Bayesian approaches to time series models.  
I did not appreciate how much functionality for this there was in the library.

ChadFulton commented:
@NickleDave thank you very much for reading the paper and for your great suggestions!  I will certainly take them on board in the my next push. 

> One point I did not comment on at a line level:
I see you wrote library and variable names in italics in most places.
I'm looking at the rendered pdf here:
https://procbuild.scipy.org/download/ChadFulton-paper-96
Is that intentional or did you mean those to appear as monospace?
(I always forget that in rst you need to use two backticks, unlike markdown that only uses one for monospace, which is why I ask.)

Thanks for noting this!

deniederhut commented:
@scoobies mark needs more review

deniederhut commented:
Thanks David!

sidsriva commented:
Hi Chad, interesting paper with a very functional library. I just wanted to quickly suggest some changes before the final review. 
And I apologize for not suggesting line changes, I read the rendered pdf available at [https://procbuild.scipy.org/download/ChadFulton-paper-96
](https://procbuild.scipy.org/download/ChadFulton-paper-96
)
(1) The equation in Autoregressive moving-average model section should have $\theta_1 \epsilon_{t-1}$ 
in place of $\theta_1 \epsilon_{t}$
(2)  In Vector autoregressive models, it will be helpful if you point out the dimensionality of matrices $\Phi$ 
(The equation does not work for arbitrary $\Phi$ 
it is a row vector of the same length as $y_t$
). You should also distinguish between the parameters  $\nu, \epsilon_t$. 
If $\nu$ 
appears just as an intercept it should be pointed out. 
(3) Same with the correlation matrices $H_t, Q_t$ 
matrices in the state space model. 
(4) I think it will be good if you specify what simulation smoother you are using, by the context I think it is the Linear Gaussian State Space model. But I think there is some benefit in highlighting it - and I apologize if I missed it. Also, you can use the definition from the Durbin and Koopman paper that you have already cited.

deniederhut commented:
@scoobies mark pending comment

deniederhut commented:
Thanks Siddhartha!

ChadFulton commented:
Thank you very much @sidsriva for the very useful suggestions.  I will incorporate them in my next push.

To make sure I'm don't miss something, I wanted to clarify your suggestions (3).  I interpret this as the good suggestion to give the dimensions of these matrices in the text.  Is that correct?

sidsriva commented:
> To make sure I'm don't miss something, I wanted to clarify your suggestions (3). I interpret this as the good suggestion to give the dimensions of these matrices in the text. Is that correct?

Yes.

deniederhut commented:
Hi @ChadFulton ! How are the revisions coming along?

ChadFulton committed:
ENH: incorporate reviewer 1 comments

ChadFulton committed:
ENH: incorporate reviewer 2 comments

ChadFulton committed:
BUG: correct the reported coverage of the credible intervals

ChadFulton commented:
Thank you again to both reviewers. I believe that I have incorporated all of the suggestions into the most recent push. In 8061a33 I also corrected the reported coverage interval (they show the range between the 10th and 90th percentiles, so they are 80% credible intervals, not 90%).

ChadFulton commented:
> (4) I think it will be good if you specify what simulation smoother you are using, by the context I think it is the Linear Gaussian State Space model. But I think there is some benefit in highlighting it - and I apologize if I missed it. Also, you can use the definition from the Durbin and Koopman paper that you have already cited.

@sidsriva this was a good suggestion, and I have clarified the two implementations that are available in Statsmodels in footnote 7.  Please let me know if there is more detail that you would find useful.

sidsriva commented:
@ChadFulton , thanks for the changes. This answers all my questions.

NickleDave reviewed (approved):
LGTM, thank you @ChadFulton

stargaser commented:
~~@NickleDave do you have any more comments on readiness of the revised version? We are collectively working to wrap these up by tomorrow.~~

Nevermind, I see you've already approved this version!

stargaser commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1093/biomet/89.3.603 is OK
- 10.1111/1467-937X.00050 is OK
- 10.1162/REST_a_00608 is OK
- 10.1162/003465398557447 is OK
- 10.7717/peerj-cs.55 is OK
- 10.25080/Majora-ebaa42b7-012 is OK
- 10.25080/Majora-92bf1922-011 is OK
- 10.7287/peerj.preprints.3190v2 is OK
- 10.21105/joss.01143 is OK
- 10.18637/jss.v076.i01 is OK
- 10.48550/arXiv.1711.10604 is OK

MISSING DOIs

- 10.1214/14-aoas788 may be a valid DOI for title: Inferring causal impact using Bayesian structural time-series models

INVALID DOIs

- 10.1504/IJMMNO.2009.03009 is INVALID
```

stargaser commented:
@ChadFulton would you please check over the one missing and one invalid DOI per the previous comment?

ChadFulton committed:
REF: Remove invalid DOI, add missing DOI

ChadFulton commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1093/biomet/89.3.603 is OK
- 10.1111/1467-937X.00050 is OK
- 10.1162/REST_a_00608 is OK
- 10.1162/003465398557447 is OK
- 10.7717/peerj-cs.55 is OK
- 10.25080/Majora-ebaa42b7-012 is OK
- 10.25080/Majora-92bf1922-011 is OK
- 10.1214/14-aoas788 is OK
- 10.7287/peerj.preprints.3190v2 is OK
- 10.21105/joss.01143 is OK
- 10.18637/jss.v076.i01 is OK
- 10.48550/arXiv.1711.10604 is OK

MISSING DOIs

- 10.1504/ijmmno.2009.030090 may be a valid DOI for title: Efficient simulation and integrated likelihood estimation in state space models

INVALID DOIs

- None
```

ChadFulton commented:
Thanks @stargaser.  I added one missing DOI, but there is still one more issue:

I'm not sure what's best to do here with respect to the remaining DOI that is currently marked as missing:

- The original DOI I used, 10.1504/IJMMNO.2009.03009 matches the link on the publisher's website: https://www.inderscienceonline.com/doi/abs/10.1504/IJMMNO.2009.03009.  However, checking the DOI at https://www.doi.org/ does indeed show that it is not valid

- Using the suggested DOI, 10.1504/ijmmno.2009.030090 is not marked as invalid by https://www.doi.org/, however it does not lead to a valid link on the publisher's website, since it attempts to go to https://www.inderscienceonline.com/doi/abs/10.1504/IJMMNO.2009.030090, which gives a 404 error.

I'm happy to leave off the DOI in this case.

stargaser commented:
Thank you for looking into the DOIs. It's fine to leave off the invalid DOI.

stargaser commented:
@scoobies mark ready
